ankisyncd-rs: add package for anki-sync-server-rs

martinetd opened the pull request:
###### Description of changes

the current ankisyncd server is obsolete and does not work with modern anki clients,
the new version of the ankisyncd python server might work but is hard to build

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
- [x] Tested, as applicable:
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

TODO:
- [x] should definitely add something in release notes -> done
- [x] do we want to keep the old package? I'm not sure new server would work with older clients, that's something to test but ENOTIME for the forseable short term future -> removed

cc @euank @Atemu (worked on anki recently) @matt-snider (old ankisyncd maintainer)
interestingly the service has no maintainer...

martinetd committed:
ankisyncd-rs: add package for anki-sync-server-rs

the current ankisyncd server is obsolete and does not work with modern anki clients,
the new version of the ankisyncd python server might work but is hard to build

Atemu commented:
This should probably have a release note.

martinetd commented:
> This should probably have a release note.

Yes, see TODO. probably should have marked this as draft, sorry doing it now.

> @martinetd I haven't been following anki development very much. Is there a discussion you can link to where this is discussed by the maintainers of anki or ankisyncd?

Haven't discussed anything, the python and rust ankisyncd are reasonably compatible if up to date so it doesn't matter which we update -- but now we're updating anki itself, we definitely should update ankisyncd, and it looked like less effort to use the rust version.

(I've been using it for a few days without problems as a datapoint, the users and data was conserved coming from an up to date python ankisyncd, so not the one in nixos. I haven't tested coming from the old python version as it doesn't work with any of my anki clients... And that is a known issue afaik)

If you'd prefer updating the python version instead I'll be happy to close this :)

euank commented:
@matt-snider 

> Is there a discussion you can link to where this is discussed by the maintainers of anki or ankisyncd?

The most relevant issue upstream I think is this: https://github.com/ankicommunity/anki-sync-server/issues/158
On the other hand, the [rust version](https://github.com/ankicommunity/anki-sync-server-rs/issues/63) got updated already.

I think that's already a quite reasonable argument for using this rust variant, despite it being a breaking change.

The other quite compelling reason for using this is that anki-sync-server is a pain for nixpkgs to package. It's many versions behind, I don't expect anyone to actually update it, and I think switching to the rust one is pragmatic in terms of our own packaging time and effort.

matt-snider commented:
@martinetd @euank I'm probably not the best person to weigh in on this, being a bit out of touch with the anki community. My main thought with linking to a discussion was just whether this was an "official" solution or at least something which has acceptance from a large portion of the community.

If I've understood your previous comments and a bit of light research, there is now an official sync server provided by Anki but it has worse backwards compatibility and is difficult to package for nix.

I think it's great to package ankisyncd-rs but I'm not so sure this should become the default implementation used in Nixpkgs given that there _is_ an official version. The anki project has hundreds of contributors and is 10 years old. This project is just a few years old and has 3 contributors (only 1 large contributor).

I would weigh those arguments against each other. Maybe we can also find some other Nix community members that use these packages to provide their opinion?

euank commented:
> there is now an official sync server provided by Anki but it has worse backwards compatibility and is difficult to package for nix.

Fortunately, we actually do end up with that packaged and working once #221229 merges!

On the rest, I agree with you. The official one is better if that's good enough for users of the module.
I personally don't run my own sync server, and so I don't have any data-point at all on whether one or the other is sufficient.

martinetd commented:
My understanding is that this is just a wrapper above the sync server code in anki itself -- the build here takes the anki sources and applies the patch in the anki-sync-server-rs which basically just makes functions accessible from outside, and then use it with a few bells and whistles.
(The server in anki still has the compat protocol at this point, but for example the sync server in anki itself requires hardcoding the users/passwords in env vars before startup, but ankisyncd (both versions) have a sqlite database with user names and hash. This is both more flexible (don't need a restart to add/modify users) and secure (don't need plain passwords))

Even without these bells and whistles, anki itself requires qt, plenty of node stuff, lame/mpv... while the sync server is mostly standalone with close to 0 dep. My server running the sync server isn't my workstation where I run the client, it doesn't make sense to have the same bundle with everything.


Honestly I don't see the need for both the python and the rust versions of the standalone server, so I don't think we need to package both, but I think we ought to package either for the reasons given above.

Given the database itself are compatible between the two so we can always switch (I've just switched from a recent python ankisyncd installed from pip to this rust package myself last week) as long as we switch the config as well, so we're not locking ourselves in anything either and we can just say ok let's go back to python later if the rust port dies; but at this point the rust server seems easier to maintain for us so I also don't see a reason not to use it given our python ankisyncd is obsolete (doesn't currently work with ankidroid nor my semi-oldish 2.1.54 on this laptop, so there's been another breakage before that 2.1.57+ link)
If someone wants to try to update the python version that has a longer history I don't mind either way, but I tried a few years ago and I'm not trying again when the rust version was so easy to do -- I've set myself as maintainer to the rust package and will be happy to keep it up to date when new releases come out.

euank commented:
You using this module and finding this rust server better than `anki --syncserver` is a plenty good argument for me.

I, personally, would be happy to approve and see this PR merged once we have the release note indicating this change is happening, for whatever my opinion is worth.

And yeah, I think the community python version should probably be dropped since I don't really expect anyone to take the time to update it, especially since you could use this version instead of spending hours of time fighting with build-systems.

martinetd commented:
- fixed the rust build, for some reason on my local branch I could use cargoDeps but got yelled at by the build system after rebasing and I hadn't noticed. Should properly build now.
- added a changelog note
- removed the old python ankisyncd and renamed ankisyncd-rs to ankisyncd so it's (almost) a drop in replacement

Sorry for double push notification I fumbled a bit with a rebase on the cargolock change.

euank commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -288,6 +288,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 - `services.dhcpcd` service now don't solicit or accept IPv6 Router Advertisements on interfaces that use static IPv6 addresses.
   If network uses both IPv6 Unique local addresses (ULA) and global IPv6 address auto-configuration with SLAAC, must add the parameter `networking.dhcpcd.IPv6rs = true;`.
 
+- The module [services.ankisyncd](#opt-services.ankisyncd.package) has been switched to [anki-sync-server-rs](https://github.com/ankicommunity/anki-sync-server-rs) from the old python version, which was difficult to update and not updated in a while, not supporting recent clients. The config file is different, but users' database is compatible so at most a rsync from clients might be needed.
```
Couple wording nits:

1. "rsync" -> "resync"
2. "which was difficult to update and not updated in a while, not supporting recent clients" -> "which was difficult to update, had not been updated in a while, and did not support recent versions of anki." (the last bit is a sentence fragment otherwise I think).

euank commented:
Running `nix run nixpkgs#nixpkgs-review -- pr 224366 --print-result` failed for me with the below:

<Details>

<Summary>nixpkgs-review output</Summary>

```
$ git worktree add /home/esk/.cache/nixpkgs-review/pr-224366/nixpkgs 328636cafa83bcdf08b2a4f4e6b7fbcbe4ec56b3
Preparing worktree (detached HEAD 328636cafa8)
Updating files: 100% (34342/34342), done.
HEAD is now at 328636cafa8 Merge pull request #224365 from fgaz/qrcp/0.10.0
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f /home/esk/.cache/nixpkgs-review/pr-224366/nixpkgs -qaP --xml --out-path --show-trace --no-allow-import-from-derivation
$ git merge --no-commit --no-ff f4b813bfaef3f3c43bd30345be2abca636a4decf
Auto-merging nixos/doc/manual/release-notes/rl-2305.section.md
Auto-merging pkgs/top-level/all-packages.nix
Automatic merge went well; stopped before committing as requested
$ nix-env --extra-experimental-features no-url-literals --option system x86_64-linux -f /home/esk/.cache/nixpkgs-review/pr-224366/nixpkgs -qaP --xml --out-path --show-trace --no-allow-import-from-derivation --meta
error:
       … while querying the derivation named 'ankisyncd-1.1.0'

       … while evaluating the attribute 'cargoDeps' of the derivation 'ankisyncd-1.1.0'

       at /home/esk/.cache/nixpkgs-review/pr-224366/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:302:7:

          301|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          302|       name =
             |       ^
          303|         let

       … while evaluating the attribute 'buildCommand' of the derivation 'cargo-vendor-dir'

       at /home/esk/.cache/nixpkgs-review/pr-224366/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:302:7:

          301|     // (lib.optionalAttrs (attrs ? name || (attrs ? pname && attrs ? version)) {
          302|       name =
             |       ^
          303|         let

       error: getting status of '/home/esk/.cache/nixpkgs-review/pr-224366/nixpkgs/pkgs/servers/ankisyncd/Cargo.lock': No such file or directory
https://github.com/NixOS/nixpkgs/pull/224366 failed to build
```

</Details>

euank commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,67 +1,59 @@
-, python3
-, anki
-}:
+{ lib, stdenvNoCC, fetchFromGitHub, rustPlatform
+, protobuf }:
```
I feel like this is short enough it reads better all on one line

martinetd replied:
I somewhat like to separate "core" dependencies to build inputs, but it doesn't really matter I guess. It's indeed short enough to fit in a line, changed.

euank commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,67 +1,59 @@
+      patch -d "$out/anki" -Np1 < "$src/anki_patch/d9d36078f17a2b4b8b44fcb802eb274911ebabe7_anki_rslib.patch"
+    '';
+  };
+in rustPlatform.buildRustPackage rec {
```
The `rec` there isn't needed (thanks [nil](https://github.com/oxalica/nil))

martinetd replied:
Thanks! I need to setup some LSP.

euank reviewed (approved):
two nits, but I don't think either of them are important.

This package runs for me, and the overall change makes sense to me!

matt-snider commented:
> If someone wants to try to update the python version that has a longer history I don't mind either way, but I tried a few years ago and I'm not trying again when the rust version was so easy to do -- I've set myself as maintainer to the rust package and will be happy to keep it up to date when new releases come out.
>
> Honestly I don't see the need for both the python and the rust versions of the standalone server, so I don't think we need to package both, but I think we ought to package either for the reasons given above.

I'm not sure if there is a miscommunication, but I'm not suggesting that we consider using the python version (ankicommunity/anki-sync-server) if you say it isn't being maintained and is difficult to package. I'm asking why we don't use the official built in version. Anki is already packaged for Nix, so it would be as simple as `${pkgs.anki}/bin/anki --syncserver`.

We have https://github.com/NixOS/nixpkgs/pull/221229 on the way, so wouldn't it make more sense to wait for that and use that instead as the new official version which is also maintained by the Anki team?

> Even without these bells and whistles, anki itself requires qt, plenty of node stuff, lame/mpv... while the sync server is mostly standalone with close to 0 dep. 

A 4000 line Cargo.lock suggests otherwise. It has plenty of dependencies, they just aren't top-level packages in Nix. Also, as commented above, anki is already packaged

> My server running the sync server isn't my workstation where I run the client, it doesn't make sense to have the same bundle with everything.

This feels like a bit of an apples to oranges comparison. My point is it just isn't a strong argument to say "I have dependencies A & B with solution 1, and X & Y with solution 2 and so solution 2 is better". Why though? Are dependencies A & B larger, i.e. you are concerned about disk space? Do they use more memory? What is the specific concern?

> You using this module and finding this rust server better than anki --syncserver is a plenty good argument for me.

I politely disagree. I want to reiterate that I'm not using this package anymore but I'm sure other people are and introducing a breaking change because one person "finds it better" isn't a strong argument.

On that note, the comment in the release notes says: "The config file is different, but users' database is compatible so at most a resync from clients might be needed."

So are you saying the user does not need to anything besides run resync from their Anki client? Or do they need to update their config? If we go with this solution I think the release note needs to be more explicit about what steps are needed. Which functionality is supported in this new server? Is it 100% compatible with the existing solution?

euank commented:
I think I understood the suggestion to use the `anki --syncserver` version, and I generally would want to use it too...

But, unfortunately, we have two constraints:

1. The existing nixos module was built with the community one, so it uses a sqlite database, allows user management, etc
2. We have one user of the module who prefers not to use the `anki --syncserver` version, and no other real data-points to work from

Because of those two factors, I think it's a harder sell to switch to `anki --syncserver` for the nixos module.

> I politely disagree. I want to reiterate that I'm not using this package anymore but I'm sure other people are and introducing a breaking change because one person "finds it better" isn't a strong argument.

I think our realistic options for people using the python version of this module at this point are:

1. Leave them running an ancient anki-syncserver version effectively forever
2. Delete the module (which will happen with 1 eventually too)
3. Switch the module to the rust version, which is kinda-sorta backwards compatible
4. Switch to the module `anki --syncserver`, which is less backwards compatible
5. Delete the module, introduce either 3 or 4 (or both) as a new module with even less expectation of even sorta backwards compatibility

I think the most compelling options are 2, 3, and 5. I see 5 as just a better version of 4, since 4 seems like it's a breaking-enough change it really does merit a new module.

It sounds like you're suggesting 4, or perhaps 5, as the ideal option, right?

I can understand that opinion, especially since upstream anki pretty much [requests us to prefer the official syncserver](https://docs.ankiweb.net/sync-server.html) in their docs ("[third party syncservers] are not recommended").

... but a preference from an actual user of the module to me seems like a stronger argument than even the upstream recommendation.

I think the difference in our opinion is in weighing the cost of a new package in nixpkgs, upstream recommendations, and our meager amount of user preference data we have. Are there other factors that make 4 or 5 more compelling to you?

euank commented:
I think I understood the suggestion to use the `anki --syncserver` version, and I generally would want to use it too...

But, unfortunately, we have two constraints:

1. The existing nixos module was built with the community one, so it uses a sqlite database, allows user management, etc
2. We have one user of the module who prefers not to use the `anki --syncserver` version, and no other real data-points to work from

Because of those two factors, I think it's a harder sell to switch to `anki --syncserver` for the nixos module.

> I politely disagree. I want to reiterate that I'm not using this package anymore but I'm sure other people are and introducing a breaking change because one person "finds it better" isn't a strong argument.

I think our realistic options for people using the python version of this module at this point are:

1. Leave them running an ancient anki-syncserver version effectively forever
2. Delete the module (which will happen with 1 eventually too)
3. Switch the module to the rust version, which is kinda-sorta backwards compatible
4. Switch to the module `anki --syncserver`, which is less backwards compatible
5. Delete the module, introduce either 3 or 4 (or both) as a new module with even less expectation of even sorta backwards compatibility

I think the most compelling options are 2, 3, and 5. I see 5 as just a better version of 4, since 4 seems like it's a breaking-enough change it really does merit a new module.

It sounds like you're suggesting 4, or perhaps 5, as the ideal option, right?

I can understand that opinion, especially since upstream anki pretty much [requests us to prefer the official syncserver](https://docs.ankiweb.net/sync-server.html) in their docs ("[third party syncservers] are not recommended").

... but a preference from an actual user of the module to me seems like a stronger argument than even the upstream recommendation.

I think the difference in our opinion is in weighing the cost of a new package in nixpkgs, upstream recommendations, and our meager amount of user preference data we have. Are there other factors that make 4 or 5 more compelling to you?

matt-snider commented:
Thank you both for your responses!

> I think our realistic options for people using the python version of this module at this point are:

It's good that you summarized this. This is taking the discussion in a more productive direction. I think the only solutions that makes sense to me are 3 ("Switch the module to the rust version, which is kinda-sorta backwards compatible") or 4 ("Switch to the module anki --syncserver, which is less backwards compatible"). Deleting the module does not seem like a sensible option to me. Switching to the mostly compatible rust version is definitely better.

I think the most compelling option might be to introduce this change with ankisyncd-rs and then hope that someone packages the official sync server as a separate module. Maybe one of you or one of the Nix `anki` maintainers are interested?

> I can understand that opinion, especially since upstream anki pretty much [requests us to prefer the official syncserver](https://docs.ankiweb.net/sync-server.html) in their docs ("[third party syncservers] are not recommended").
> 
> ... but a preference from an actual user of the module to me seems like a stronger argument than even the upstream recommendation.

For a distribution the size of Nix, I don't think the opinion of a single user is a compelling argument. But I think this one is:

> No, as said above the user configuration is wildly different (anki itself only supports SYNC_USERx env variables with clear text password, while ankisyncd use a database where passwords are stored as hash (salted sha256). Even if that was the only difference, that'd be enough argument for me.

It's unfortunate that this module didn't abstract the credentials as part of the config, because then we might have had a seamless change. But I agree with you, if we made that switch, users would need to update the config to set their password. I think most of the other configuration changes would be seamless though (the module would handle it).

> Yes, yes, sure I didn't word it properly. Nevermind that anki's Cargo.lock is 4849 lines [...] let's just compare the closure size of the new anki and this anki-sync-server-rs output derivations

Fair enough. 1.5 GB vs 42 MB is large enough of a difference to be relevant to this discussion.

> And I've now officially spent more time checking facts and replying here than time it took me to make this package and update the module, I guess I should have just not bothered and I can keep the module in my overlay...

I'm sorry that this has been a frustrating experience for you. I don't have a major stake in this, but I use and benefit from NixOS immensely and try to contribute where I can (which isn't much). My intention was not to make this difficult, but to make the "right" choice. NixOS is a big distribution and used by many people and that community counts on contributors like you to make the right decisions. A few extra hours of discussion and review time can save a lot of time when multiplied across the large and quickly growing user base.

I appreciate you and @euank taking the time to discuss this. As I said above, I think the best option is to move forward with this change for the immediate benefit to users of this module and then in the future introduce a separate module for the official client for those who would prefer that.

Tagging @oxij and @Profpatsch as they are the `anki` maintainers. I think you both could skim the last few comments (likely not necessary to read the whole discussion) and then add input if you have any.

matt-snider commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,67 +1,58 @@
-    homepage = "https://github.com/ankicommunity/anki-sync-server";
-    license = licenses.agpl3Only;
-    platforms = platforms.linux;
+    description = "Standalone anki sync server";
```
Maybe a bit more clear with: Standalone unofficial anki sync server

matt-snider replied:
Maybe a bit more clear with: Standalone **unofficial** anki sync server

martinetd committed:
replace ankisyncd with ankisyncd-rs

- remove old ankisyncd python package
- rename new rust package so it's a drop in replacement

martinetd commented:
Sorry I hadn't noticed your reply (and been rather busy so didn't check).

I've just:
 - rebase, resolving the merge conflict (python version using anki submodule, irrelevant for this PR)
 - moved the release notes section from 2305 to 2311 -- that'll give people plenty of time to test on unstable, better than getting this in just before the release anyway.
 - updated the description wording as suggested, thanks!
 - updated version to 1.1.3 which was released last month; unfortunately the release notes aren't quite clear what these bring... but upon checking there is exactly no change for us, it's only updates to the build script (changes such as using musl instead of glibc in their binary release); this update is just so we won't get a semi-automatic update PR immediately after this gets merged.

euank reviewed (approved):
This still looks good to me after the rebase / small updates described above, thanks!

martinetd commented:
@ofborg build ankisyncd

(not sure what that internal error is about, but let's try to have it build again...)

Atemu commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,68 +1,58 @@
 
-python3.pkgs.buildPythonApplication rec {
+let
   pname = "ankisyncd";
```
Perhaps

```suggestion
  pname = "ankisyncd-rs";
```

Atemu replied:
Up to you though.

martinetd replied:
the original name for either the old python version and new rust ones are anki-sync-server and anki-sync-server-rs which aren't the same, and both provide the 'ankisyncd' binary, so I'm not sure honestly...

Could you remind me what pname is used for? nix shell and similar will use whatever is defined in all-package's toplevel, so pname is for listing ouputs?

Atemu replied:
It should be called are anki-sync-server and are anki-sync-server-rs respectively then.

pname is purely for informational purposes; for us humans to interpret.

Atemu commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,68 +1,58 @@
+  # We apparently cannot do that in postPatch or equivalent in a
+  # buildRustPackage derivation because the cargo deps are processed
+  # first, there might be a better way of doing this? But it works
+  ankiSrc = fetchFromGitHub {
+    owner = "ankitects";
+    repo = "anki";
+    rev = "2.1.60";
+    hash = "sha256-hNrf6asxF7r7QK2XO150yiRjyHAYKN8OFCFYX0SAiwA=";
     fetchSubmodules = true;
   };
```
Could you add reasoning for why this is pinned rather than just taking `anki.src`?

martinetd replied:
this has a few patches on top of anki, so needs a specific version to build -- if we update anki this will need to rebase the patches to build with the new version as well, and I'm not sure how timely upstream updates are in that respect (the sync protocol doesn't change all the time, so there should be no problem running with a slightly older anki version for its protocol definition, but I just haven't used this for long enough to tell)

I'll add a comment explaining this.

Atemu commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,68 +1,58 @@
-    decorator
-    requests
-  ];
+  src = stdenvNoCC.mkDerivation {
```
runCommand

Atemu commented on nixos/modules/services/misc/ankisyncd.nix:
```diff
@@ -9,22 +9,16 @@ let
-      data_root = stateDir;
-      auth_db_path = authDbPath;
-      session_db_path = sessionDbPath;
-
-      base_url = "/sync/";
-      base_media_url = "/msync/";
     };
-  });
+    paths.root_dir = stateDir;
+    # encryption.ssl_enable / cert_file / key_file
+  };
```
Perhaps be nice to have an RFC42-like settings attr for additional settings.

martinetd replied:
I focused on migrating equivalent functionality but I guess that might be useful for someone at some point, yes... And it's trivial now we use toml.generate. I guess I can add that tonight.

At this rate though I'd say I want some nixos/test/ankisync.nix first and I'm not sure that'll end well, but I'll leave that for a much later future me; I have no idea how to start the anki client interactively to test sync and don't want to get started with GUI testing :P

Atemu replied:
Some basic smoke tests would already be pretty good; whether it runs at all or can answer some basic requests.

martinetd replied:
I had a quick look at the anki protocol and it's a real monster (some binary data sent as post to get a key, then some more binary data including the key to get stats like last modfcations etc.

Can probably just replay something straight out of tcpdump through curl but I have no idea how fragile that'll be, I'll just slack for now.

Also didn't do free form yet -- I didn't remember how to integrate existing felds in the free form nicely and ran out of time for today; might do it later but imo it's orthogonal enough that we don't need to do it now (perfect is the enemy of good etc); I'd rather do in a later PR to keep things separate.

Atemu replied:
Either mkMerge (mkDefault the defaults) or use the built-in update operator (`//`). First would probably be cleaner.

Not a hard requirement; I won't block on this.

Atemu commented on pkgs/servers/ankisyncd/default.nix:
```diff
@@ -1,68 +1,52 @@
-, fetchFromGitHub
-, python3
-}:
+{ lib, runCommand, fetchFromGitHub, rustPlatform, protobuf }:
```
The file should probably be moved aswell.

Atemu replied:
Yes. pname should almost always match the directory name.

I'd do pname and file name as anki-sync-server and the NixOS option as ankisyncd. Then add a package alias `ankisyncd = anki-sync-server;` and also not forget `meta.mainProgram = "ankisyncd";`.

martinetd replied:
Hmm, I'm just really not convinced people would look for anki-sync-server at all -- that's the github repo name alright, but I've always seen it written ankisyncd everywhere else, and I think it is easier to find if package/service directory names match.

I didn't realize pname/directory name should match so I've just reverted pname back to "ankiysncd", and I'd rather just keep it like this for now: the old package's repo was also anki-sync-server packaged as ankisyncd, this just replaces the outdated python implem with the more up to date rust one.

Sorry for the extra round-trips!

Atemu reviewed (approved):
Diff LGTM

martinetd committed:
ankisyncd: 1.1.3 -> 1.1.4

martinetd commented:
- rebased (relase note conflict)
- upgraded ankisyncd to 1.1.4; I've kept it as a separate commit because it's a pain with the rename and I kind of gave up; if someone tells me it'll get merged I'll be happy to squash the upgrade in the commit that introduces the package.

I've built it but not tested yet (system rebuilding with it); will confirm tomorrow but not expecting any problem. release notes says it allows modifying users dynamically so I assume it was slurping the sqlite db on service start previously, and now queries it when appropriate... If you don't hear from me in ~8 hours please assume the service works (that or a bus ran into my bedroom)

martinetd committed:
nixos/ankisyncd: use ankisyncd-rs instead of the old python one

Unfortunately the config is not compatible; data itself looks like it was compatible
from an up to date python ankisyncd but I wouldn't assume anything with the older
service (which didn't work for me with either ankidroid or recent desktop version)

martinetd commented:
(rebased again to fix release note merge conflict again; and I can confirm the minor update works without problem by now...)

Atemu reviewed (approved):
This won't get any better by just sitting here, this is ready for unstable and I trust @martinetd to fix issues that crop up.

Could you make an announcement in https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/?

martinetd commented:
Thank you!

> Could you make an announcement in [discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574](https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/)?

Done

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/27

dae commented:
A heads up that the developer of this appears to have abandoned it, and it will fail to work with the next Anki release due to minor protocol changes.

martinetd commented:
> A heads up that the developer of this appears to have abandoned it, and it will fail to work with the next Anki release due to minor protocol changes.

Thanks for the heads up! That's very nice of you.
I was also just pointed out that the project had been abandoned on the nixos forum a couple of days ago, but if the protocol changes this soon this will be more interesting than I thought...

Would you be able to confirm a bit more precisely what protocol change this is about? A quick search on anki commits only turned up https://github.com/ankitects/anki/pull/2329 but I believe the anki-sync-server-rs is already on top of this, and I didn't see any other commit with 'protocol' in the message either merged recently or in recent PRs.


afaiu this (anki-sync-server-rs) appears to be using anki itself to do a bulk of the work, so if just rebasing ought to suffice then I'll be happy to keep rebasing the code for the forseeable future, but it'd help to understand what changed exactly -- I'm sure you've taken compatibility with older clients in consideration but this has been a problem in the past so I'd like to check that too.
(and ultimately I guess even I will finally give in an use the built-in server, but perhaps it'd be acceptable to make it accept the same sqlite format as the anki-sync-server* implementations have been using first?)

dae commented:
Typically, a newer server will work with older clients, but newer clients will not work with an older server. The change I was referring to is https://github.com/ankitects/anki/pull/2647.

I'm afraid I'm [not keen](https://docs.ankiweb.net/sync-server.html#contributing-changes) on adding a dependency on SQLite. But one of the issues raised above has been addressed - since 2.1.66, the built-in sync server can be built [without Python](https://docs.ankiweb.net/sync-server.html#with-cargo).

martinetd commented:
Thanks for the links and confirmation!

My incomplete understanding makes me think a quick rebase should work and I've been able to build just fine, so I'll test this over the next weekend and open a PR to anki-sync-server-rs to update / update this package here after confirming it works.

> But one of the issues raised above has been addressed - since 2.1.66, the built-in sync server can be built [without Python](https://docs.ankiweb.net/sync-server.html#with-cargo).

This is great -- if this had been around back in May I might have given up on compatibility with the old modules; I'd be much less reluctant to use this server than the full bundle.

I'm honestly split 50/50 between how much pain it'll be to update in the long run and wanting to keep compatibility with the original anki-sync-server module sqlite db... As long as the anki-sync-server-rs stays this simple to update I guess I can take up ownership for a while, and we can break compat when that ship sails?


(BTW I'll be interested in further notices when the protocol gets updated again, but it'd be selfish to ask you to ping me everytime -- if you have a preference for this (issues on the anki-sync-server-rs project?) I'll be happy to add a watch somewhere)

dae commented:
> if this had been around back in May

It wasn't [particularly complicated](https://github.com/ankitects/anki/blob/main/rslib/sync/main.rs). :-)

I'm afraid I can't guarantee I'll remember to update you every time there's a breaking change, so if you wish to keep using anki-sync-server-rs, I'd recommend you check it's still functional after each Anki release. There's at least one more minor protocol change that's likely to land in the next few releases, related to config handling.

matt-snider commented:
@martinetd I was pushing for the switch to the official version before because I was concerned about the longevity of the anki-sync-server-rs project.

Now that it is abandoned I feel it's one more argument in favour of switching to using the official version. Otherwise it seems like you are taking on responsibility for maintaining that project going forward which seems like it might be more work long-term than making the switch.

From @euank 's comment in this thread:

> Fortunately, we actually do end up with that (`anki --syncserver`) packaged and working once #221229 merges!

And @dae 's comment in this thread:

> But one of the issues raised above has been addressed - since 2.1.66, the built-in sync server can be built [without Python](https://docs.ankiweb.net/sync-server.html#with-cargo).

I wonder if there have been sufficient simplifications to building and using the built-in sync server that you might also agree that it's a better option at this point. Looking to hear your opinion on that as I'm kind of far removed from using sync functionality at this point.

martinetd commented:
Yes, I agree the simplifications on anki side are great -- building a standalone anki-sync-server package from anki sources ought to be very simple and would get a working server without all the anki dependencies I was complaining about earlier this year.
As said previously, if this had been around when I did this work I probably wouldn't have spent so much time arguing for anki-sync-server-rs, and just packaged that instead... I have no problem using the to-be anki server package wise.

Now we're here though it's a bit of a sunk time fallacy problem: we have something working that keeps module compatibility through `services.ankisyncd`, and as I looked yesterday it really isn't much work to keep up (unlike the python implementation, since this server uses the main anki server's code it really just is a rebase -- I was able to get the new version to build in minutes); that's the main reason I wasn't worried about maintenance earlier this year, and as long as it only breaks a couple of times a year I really don't see much of a problem doing that once in a while.
(Not tested yet but I don't see why it wouldn't work -- probably would want to invest a bit of time setting up a nixos test so we don't rely on @dae's good will to reach out everytime though, and that'll take some time, but I think it's something we ought to do regardless of which server is used and the hard part is on anki side to drive a couple of syncs through the qt interface...)


Basically it boils down to a bit of my time vs. compatibility with existing setups, which we have no way of knowing how many users would be impacted (just because the old python sync server wouldn't be working with new client doesn't mean people weren't using it with older clients... I was one of them)
OTOH while it's been 4 months there haven't been a nixos release yet (this is all planned for 23.11), so while we're breaking old clients it might be better to break the old sqlite db as well...

But we really need to make sure the data itself (cards and media) is compatible otherwise it's better to do this in two steps (1/ keep data on server (nixos module config unchanged, anki-sync-server-python > anki-sync-server-rs), upgrade clients; 2/ keep clients, switch server to anki main server and resync from client to preserve data).
If data is compatible this isn't as much of a problem, but it'll still require fiddling with the nixos config and upgrading the clients at the same time for users so slightly more chance of causing a few headaches.

Since I haven't verified that I cannot say anything really, but I'm probably am too involved in this to make a fair decision anyway.

martinetd commented:
I haven't taken the time to build anki's master branch, but at least my rebased ankisyncd still works with current anki releases so I've opened https://github.com/ankicommunity/anki-sync-server-rs/pull/76/ to prepare this side.

Unfortunately I can't get it to build in nixpkgs (tested building manually)... And anki will have the same problem whenever someone tries to update it, because our cargo vendoring script makes a link for each crate we depend on as $crate-$version and the latest anki depends on two different csv-core v0.1.10:
```
├── csv v1.1.6 (https://github.com/ankitects/rust-csv.git?rev=1c9d3aab6f79a7d815c69f925a46a4590c115f90#1c9d3aab)
│   ├── csv-core v0.1.10 (https://github.com/ankitects/rust-csv.git?rev=1c9d3aab6f79a7d815c69f925a46a4590c115f90#1c9d3aab)
├── fsrs-optimizer v0.1.0 (https://github.com/open-spaced-repetition/fsrs-optimizer-rs?rev=e0b15cce555a94de6fdaa4bf1e096d19704a397d#e0b15cce)
│   ├── burn v0.9.0 (https://github.com/burn-rs/burn.git?rev=36446e8d35694a9158f97e85e44b84544b8c4afb#36446e8d)
│   │   ├── burn-core v0.9.0 (https://github.com/burn-rs/burn.git?rev=36446e8d35694a9158f97e85e44b84544b8c4afb#36446e8d)
│   │   │   ├── burn-dataset v0.9.0 (https://github.com/burn-rs/burn.git?rev=36446e8d35694a9158f97e85e44b84544b8c4afb#36446e8d)
│   │   │   │   ├── csv v1.2.2
│   │   │   │   │   ├── csv-core v0.1.10
```

So we'll either need to patch anki to depend on another version of csv (had a look at the anki patch and we can't drop it, but we can just have it change the csv-core version so it doesn't collide), or teach buildRustPackages' cargo-vendor-dir to deal with duplicate crates name/version pairs if one is a git repo...

So, yeah, this is more work than expected but work we'll have to do for anki at some point anyway so my opinion hasn't really changed; I'm done for this weekend but I'll eventually have another look if no-one beats me to it.
